File: crates/corral-core/src/amount.rs
//! `TokenAmount` — the only representation of a token quantity in the entire
//! system (CLAUDE.md §2.10). Base units, checked arithmetic, decimal-string
//! wire format. A silent overflow in a spending limit is the worst bug this
//! system can have, so overflow is unrepresentable rather than discouraged.

use alloy_primitives::U256;
use core::fmt;
use serde::{de, Deserialize, Deserializer, Serialize, Serializer};

/// A token amount in base units.
///
/// Deliberately NOT convertible to or from floats or primitive integers
/// without an explicit call, and deliberately without `Add`/`Sub`/`Mul`
/// impls: every arithmetic step must go through a `checked_*` method and
/// handle the `None` case.
///
/// Wire format is a decimal string (`"500000000"`), matching Postgres
/// `numeric(78,0)`. Hex, exponents, signs, separators and bare JSON numbers
/// are rejected on input.
#[derive(Debug, Clone, Copy, Default, PartialEq, Eq, PartialOrd, Ord, Hash)]
pub struct TokenAmount(U256);

impl TokenAmount {
    pub const ZERO: Self = Self(U256::ZERO);

    /// Wrap a raw base-unit value. The only way in.
    pub const fn from_base_units(v: U256) -> Self {
        Self(v)
    }

    /// The raw base-unit value. The only way out.
    pub const fn get(self) -> U256 {
        self.0
    }

    pub fn is_zero(self) -> bool {
        self.0.is_zero()
    }

    /// Checked addition: `None` on overflow. There is intentionally no
    /// unchecked, wrapping, or saturating alternative.
    #[must_use = "a None here means an overflow was prevented — it must be handled"]
    pub fn checked_add(self, other: Self) -> Option<Self> {
        self.0.checked_add(other.0).map(Self)
    }

    /// Checked subtraction: `None` on underflow.
    #[must_use = "a None here means an underflow was prevented — it must be handled"]
    pub fn checked_sub(self, other: Self) -> Option<Self> {
        self.0.checked_sub(other.0).map(Self)
    }
}

impl fmt::Display for TokenAmount {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        // U256's Display is decimal; this is also the wire format.
        write!(f, "{}", self.0)
    }
}

impl Serialize for TokenAmount {
    fn serialize<S: Serializer>(&self, serializer: S) -> Result<S::Ok, S::Error> {
        serializer.serialize_str(&self.0.to_string())
    }
}

impl<'de> Deserialize<'de> for TokenAmount {
    fn deserialize<D: Deserializer<'de>>(deserializer: D) -> Result<Self, D::Error> {
        struct V;
        impl de::Visitor<'_> for V {
            type Value = TokenAmount;

            fn expecting(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
                f.write_str("a base-unit token amount as a decimal string")
            }

            fn visit_str<E: de::Error>(self, s: &str) -> Result<TokenAmount, E> {
                // Strictly ASCII digits: no sign, no hex/binary prefix, no
                // exponent, no separators, no whitespace, not empty. U256's
                // own FromStr would accept "0x…" — too permissive for money.
                if s.is_empty() || !s.bytes().all(|c| c.is_ascii_digit()) {
                    return Err(E::invalid_value(de::Unexpected::Str(s), &self));
                }
                U256::from_str_radix(s, 10)
                    .map(TokenAmount)
                    .map_err(|_| E::custom("token amount exceeds U256::MAX"))
            }
        }
        deserializer.deserialize_str(V)
    }
}

File: packages/core/src/amount.ts
/**
 * `TokenAmount` — the only representation of a token quantity in the entire
 * system (CLAUDE.md §2.10). Base units, checked arithmetic, decimal-string
 * wire format. A silent overflow in a spending limit is the worst bug this
 * system can have.
 *
 * Parity port of `crates/corral-core/src/amount.rs`. Rust's `U256` newtype
 * made misuse uncompilable; in TS the substitutes are: a branded `bigint`
 * (plain bigints and numbers don't typecheck where a TokenAmount is
 * required), range-checked constructors, checked-only arithmetic helpers,
 * and a wire schema that accepts nothing but a pure decimal string.
 */

import { z } from "zod";

declare const TokenAmountBrand: unique symbol;

/**
 * A token amount in base units: a `bigint` proven to lie in `[0, 2^256 - 1]`.
 *
 * There are deliberately no arithmetic operators for this type — native
 * `+`/`-` on the underlying bigints would bypass the range guarantee. Every
 * arithmetic step goes through `checkedAdd`/`checkedSub` and must handle
 * `null`.
 */
export type TokenAmount = bigint & { readonly [TokenAmountBrand]: "TokenAmount" };

/** The EVM word ceiling: 2^256 − 1. */
export const U256_MAX: bigint = (1n << 256n) - 1n;

export const TOKEN_AMOUNT_ZERO: TokenAmount = 0n as TokenAmount;

/**
 * Wrap a raw base-unit value. The only way in. Throws `RangeError` outside
 * `[0, U256_MAX]` — negative and oversized amounts are unrepresentable, not
 * merely invalid.
 */
export function tokenAmount(v: bigint): TokenAmount {
  if (v < 0n || v > U256_MAX) {
    throw new RangeError(`token amount out of range: ${v.toString(10)}`);
  }
  return v as TokenAmount;
}

export function isZero(a: TokenAmount): boolean {
  return a === 0n;
}

/**
 * Checked addition: `null` on overflow past `U256_MAX`. There is
 * intentionally no unchecked, wrapping, or saturating alternative.
 * A `null` here means an overflow was prevented — it must be handled.
 */
export function checkedAdd(a: TokenAmount, b: TokenAmount): TokenAmount | null {
  const sum = a + b;
  return sum > U256_MAX ? null : (sum as TokenAmount);
}

/** Checked subtraction: `null` on underflow below zero. */
export function checkedSub(a: TokenAmount, b: TokenAmount): TokenAmount | null {
  return a < b ? null : ((a - b) as TokenAmount);
}

/** Wire form: pure decimal string, matching Postgres `numeric(78,0)`. */
export function formatTokenAmount(a: TokenAmount): string {
  return a.toString(10);
}

/**
 * Human display of a base-unit amount at a given token scale, by integer
 * math only — `Number(amount) / 10 ** decimals` loses digits above 2^53 and
 * an 18-decimal balance passes that at 0.009 ETH.
 *
 * Display only. Nothing derived from this value may be fed back into a
 * policy, a plan, or an amount: the wire form is always base units
 * (`formatTokenAmount`).
 */
export function formatUnits(a: TokenAmount, decimals: number): string {
  if (!Number.isInteger(decimals) || decimals < 0 || decimals > 77) {
    throw new RangeError(`decimals out of range: ${String(decimals)}`);
  }
  if (decimals === 0) return a.toString(10);
  const base = 10n ** BigInt(decimals);
  const whole = (a / base).toString(10);
  const frac = (a % base).toString(10).padStart(decimals, "0").replace(/0+$/, "");
  return frac === "" ? whole : `${whole}.${frac}`;
}

const DECIMAL_ONLY = /^[0-9]+$/;

/**
 * The wire schema. Strictly a string of ASCII digits: no sign, no hex or
 * binary prefix, no exponent, no separators, no whitespace, not empty, and
 * never a bare JSON number (ambiguous precision). Validated *before*
 * `BigInt()` — `BigInt(" 1")` and `BigInt("")` would silently accept.
 */
export const TokenAmountSchema: z.ZodType<TokenAmount, string> = z
  .string()
  .regex(DECIMAL_ONLY, "token amount must be a base-unit decimal string")
  .transform((s, ctx) => {
    const v = BigInt(s);
    if (v > U256_MAX) {
      ctx.addIssue({ code: "custom", message: "token amount exceeds U256_MAX" });
      return z.NEVER;
    }
    return v as TokenAmount;
  });
